### ISSUE TITLE
Http client is deadlocked when trying to send a request during pool shutdown

### ISSUE THREAD

**Issue by e-orz** [akka/akka-http#303]
I have several cases in which the client is deadlocked. The first log print appears but no log print after the request exists (the request was performed using .via(Http().superpool()). The whole stream is stuck in the Http stage. See full example below.
1. Normal stream operation:
   A complex stream (I couldn't reproduce it with a small one) that used UnfoldFlow (from contrib) to send a request every 30 seconds. Looking at the logs it seems that the connections are closing down (DEBUG level - not an issue) and when the last connection is closed the whole pool is closed. According to the logs the request is sent and a split second later the pool is closed. The stream is still alive but stuck. It seems that setting the delay to 40 seconds solves the issue (still needs more investigation)
2. The server closes the TCP (reset by peer) - reproducer below
   In this case the server sent the response (large (several MB)) and during consuming the entity's databytes the client got reset by peer. Trying to retry the request (using Retry from contrib) caused all the stream to get stuck. Looking at Wireshark, the request was never sent.
   Again, also in this case setting the delay to 40 seconds solves the issue (the request is sent only when the pool is already properly closed)

reproducer:
(set the uri to a server with large response that closes the connection during the send)

``` scala
object Main extends App with LazyLogging {
  implicit val system = ActorSystem("stream-dc")
  implicit val mat = ActorMaterializer(ActorMaterializerSettings(system))

  val initialCount = 1000
  val theUri = "..."
  case object Tick

  val getFlow = Flow[(Future[Tick.type], Int)]
    .mapAsync(1) { case (input, state) => input.map(_ => state) }
    .map(count => HttpRequest(uri = theUri) -> count)
    .log("before http", _ => "sending request")
    .via(Http().superPool())
    .log("after http", _ => "response got")
    .flatMapConcat {
      case (Success(HttpResponse(status, header, entity, protocol)), count) => {
        entity.dataBytes
          .via(Framing.delimiter(ByteString("\n"), maximumFrameLength = 1024))
          .fold(ByteString(""))((total, bs) => total ++ bs)
          .map(body => Try(body) -> count)
          .recover { case e =>
            logger.warn("entity databytes failed", e)
            Failure[ByteString](e) -> count
          }
      }
      case (Failure(ex), count) => {
        logger.error("error occurred:", ex)
        Source.single(Failure[ByteString](ex) -> count)
      }
    }
  val getWithRetry = Retry(getFlow) {
    case 0 => None
    case count => {
      logger.info("going to retry")
      Some(Future.successful(Tick) -> (count - 1)) // bad one http stuck
//      Some(akka.pattern.after(40.seconds, system.scheduler)(Future.successful(Tick)) -> (count - 1)) // good one
    }
  }
  val doneFut = Source.single(Future.successful(Tick) -> initialCount).via(getWithRetry).runForeach(element => logger.info(s"$element"))
  doneFut.onComplete {
    case Success(_) => logger.info("successfully finished")
    case Failure(ex) => logger.error("finished with error:", ex)
  }
}
```

sbt:

``` sbt
name := "AkkaHttpDeadlock"

version := "1.0"

scalaVersion := "2.11.8"

libraryDependencies ++= Seq(
  "com.typesafe.akka" %% "akka-stream" % "2.4.10",
  "com.typesafe.akka" %% "akka-http-core" % "2.4.10",
  "com.typesafe.akka" %% "akka-stream-contrib" % "0.3",
  "com.typesafe.akka" %% "akka-slf4j" % "2.4.10",
  "ch.qos.logback" %  "logback-classic" % "1.1.7",
  "com.typesafe.scala-logging" %% "scala-logging" % "3.4.0"
)
```

application.conf

```
akka {
  log-dead-letters = 2
  log-dead-letters-during-shutdown = on
  #  log-config-on-start = on
  #  jvm-exit-on-fatal-error = on

  loggers = ["akka.event.slf4j.Slf4jLogger"]
  loglevel = "DEBUG"
  logging-filter = "akka.event.slf4j.Slf4jLoggingFilter"

  http.host-connection-pool {
    max-connections = 10
    max-open-requests = 128
    client {
      parsing.max-content-length = 360M
    }
  }
}
```

**Comment by jrudolph** [comment]
Is that still an issue for you? We have fixed a few pool bugs in the meantime. Could you recheck with the latest akka-http (10.0.4 at the time of writing)?

**Comment by e-orz** [comment]
Thanks for your reply, I will try to reproduce it and will update you (in the meantime for our production as a workaround I set a timeout of 40 seconds)

**Comment by jrudolph** [comment]
Great, thanks for the update, @e-orz.

**Comment by e-orz** [comment]
Well done!
I reproduced the bug (using the reproducer below) using the old version. When I changed the version the issue is solved.

Please don't close this issue just yet. I want also to check the first scenario. It's a longer check and I hope that I will get to it next week.

**Comment by e-orz** [comment]
Hi,
Sorry it took time.
I run again the first scenario with the new version (10.0.5) and it doesn't happen any more.
Thanks for the fix and for a great library, please close this issue.

Eli

**Comment by jrudolph** [comment]
Ok, great. Thanks for checking @e-orz. Let's hope the problem is gone for good.

**Comment by jrudolph** [comment]
Could have been related to the issue reported and fixed in #1245.

**Comment by jrudolph** [comment]
It's hard to say for sure, the 30 seconds in your original description that coincides with the default `host-connection-pool.idle-timeout` would point indeed in the direction of #1245.
